### ISSUE TITLE
Add option to disable automatic decompression

### ISSUE THREAD

**Issue by ypoluektovich** [sagebind/isahc#227]
Currently, when making a request, curl's automatic response decoding is enabled unconditionally (in https://github.com/sagebind/isahc/blob/1ac051263b0a006d792bcf0310ac41fee2cda78f/src/client.rs#L940-L947). Thus, when e. g. a gzipped response is received, the Content-Length header specifies the compressed size, but the data you can get from the body is uncompressed — and thus, typically, longer, though there is no way to know *how much* longer.

**Comment by sagebind** [comment]
You can manually disable this by setting a custom `Accept-Encoding` header on your request (or as a default header on an `HttpClient`), though I agree that we could offer a more ergonomic way of doing this.

**Comment by ypoluektovich** [comment]
I could, but that would force *me* (or, since I'm actually making a proxy, my client) to "start" the negotiation process. So, what should I choose?
- I could use `*/*`, but then the server can default to something obscure that my agent does not support.
- Or I could use some safe set of encodings, like `gzip, deflate`, but then when my agent (e. g. curl) adds support for a new better algorithm (e. g. brotli), I won't get the benefit of that without reconfiguring.

**Comment by sagebind** [comment]
Ah, I see now! You'd still like to _ask for_ gzipped response encoding (or rather, your downstream client might), but you just don't want it to be decoded and instead want to pass it downstream for _them_ to decode. Makes sense.

I'm envisioning an API that looks something like this:

```rust
let client = HttpClient::builder()
    // This would be the default behavior, like today.
    .automatic_decompression(Decompress::all_supported())
    // Don't decompress any content encodings.
    .automatic_decompression(Decompress::none())
    // Only decompress the following encoding names.
    .automatic_decompression(Decompress::only("gzip", "deflate"))
```

That way, you can distinguish between which content encodings you are asking from the server vs. which ones to automatically decompress.

**Comment by ypoluektovich** [comment]
Yep! That would be perfect.

Do you have any estimate on when this feature could be added?

**Comment by sagebind** [comment]
I can't give any promises as Isahc work is dependent on my available spare time, but I can definitely aim for making something like this available in a release this week.

**Comment by sagebind** [comment]
OK, so digging deeper into this, it appears that `CURLOPT_ACCEPT_ENCODING` doesn't actually work the way that I thought it did when I proposed that API. As far as automatic decompression goes, libcurl only has two modes: decompress automatically, or don't. The list of encodings given for `CURLOPT_ACCEPT_ENCODING` is only used to set the `Accept-Encoding` header, it doesn't actually control _which_ encodings libcurl will decompress.

To that end, it looks like we'll have to go for a more minimal API:

```rust
let client = HttpClient::builder()
    // This would be the default behavior, like today.
    .automatic_decompression(true)
    // Don't decompress any content encodings.
    .automatic_decompression(false)
    .build()?;
```

**Comment by sagebind** [comment]
This is now available in the newly-minted 0.9.10 release!

**Comment by ypoluektovich** [comment]
Thanks!
